src/core/control/Control.cpp

```diff
@@ -646,21 +650,16 @@ void Control::actionPerformed(ActionType type, ActionGroup group, GtkToolButton*
             }
             break;
         case ACTION_SETSQUARE:
-            if (!this->win->getXournal()->getSetsquareView()) {
-                // bring up setsquare in page center
-                auto setsquare = std::make_unique<Setsquare>();
-                auto view = win->getXournal()->getViewFor(getCurrentPageNo());
-                std::unique_ptr<SetsquareView> setsquareView = std::make_unique<SetsquareView>(view, setsquare);
-                if (!view) {
-                    setsquareView.reset(nullptr);
-                }
-                this->win->getXournal()->setSetsquareView(std::move(setsquareView));
-                fireActionSelected(GROUP_SETSQUARE, ACTION_SETSQUARE);
+            if (auto xournal = this->win->getXournal();
+                !this->geometryToolController ||
+                this->geometryToolController->getType() != GeometryToolType::SETSQUARE) {
+                resetGeometryTool();
+                makeGeometryTool(GeometryToolType::SETSQUARE);
+                xournal->getViewFor(getCurrentPageNo())->rerenderPage();
```

Comment:
I removed the page rendering and it still works fine. I had to pass the tool range of the setsquare as the last repaint range to the geometry tool constructor, since otherwise there were artifacts when translating the setsquare using the arrow keys.